src/server/src/main/java/io/cassandrareaper/storage/CassandraStorage.java

```diff
@@ -500,99 +499,102 @@ public boolean isStorageConnected() {
               ? objectMapper.readValue(row.getString("properties"), ClusterProperties.class)
               : ClusterProperties.builder().withJmxPort(Cluster.DEFAULT_JMX_PORT).build();
-      return row != null
-          ? Optional.ofNullable(
-              new Cluster(
-                  row.getString("name"),
-                  Optional.ofNullable(row.getString("partitioner")),
-                  row.getSet("seed_hosts", String.class),
-                  properties))
-          : Optional.empty();
-    } catch (RuntimeException | IOException e) {
-      LOG.error("Failed parsing cluster information from the database entry", e);
-      throw new ReaperException(e);
+
+        Cluster.Builder builder = Cluster.builder()
+              .withName(row.getString("name"))
+              .withSeedHosts(row.getSet("seed_hosts", String.class))
+              .withJmxPort(properties.getJmxPort())
+              .withState(null != row.getString("state")
+                  ? Cluster.State.valueOf(row.getString("state"))
+                  : Cluster.State.UNKNOWN)
+              .withLastContact(row.get("last_contact", LocalDate.class));
+
+        if (null != row.getString("partitioner")) {
+          builder = builder.withPartitioner(row.getString("partitioner"));
+        }
+        return builder.build();
+      } catch (IOException e) {
+        LOG.error("Failed parsing cluster information from the database entry", e);
+        throw new IllegalStateException(e);
+      }
     }
+    throw new IllegalArgumentException("no such cluster: " + clusterName);
   }
 
   @Override
-  public Optional<Cluster> deleteCluster(String clusterName) {
-    assert getRepairSchedulesForCluster(clusterName).isEmpty()
-        : StringUtils.join(getRepairSchedulesForCluster(clusterName));
+  public Cluster deleteCluster(String clusterName) {
 
-    assert getRepairRunsForCluster(clusterName, Optional.of(Integer.MAX_VALUE)).isEmpty()
-        : StringUtils.join(getRepairRunsForCluster(clusterName, Optional.of(Integer.MAX_VALUE)));
+    assert getRepairSchedulesForCluster(clusterName).stream()
+          .anyMatch(schedule -> RepairSchedule.State.ACTIVE == schedule.getState())
+        : "Cluster has active schedules " + StringUtils.join(getRepairSchedulesForCluster(clusterName));
 
-    Statement stmt = new SimpleStatement(SELECT_REPAIR_UNIT);
-    stmt.setIdempotent(Boolean.TRUE);
-    ResultSet results = session.execute(stmt);
-    for (Row row : results) {
-      if (row.getString("cluster_name").equals(clusterName)) {
-        UUID id = row.getUUID("id");
-        assert getRepairRunsForUnit(id).isEmpty() : StringUtils.join(getRepairRunsForUnit(id));
-        session.executeAsync(deleteRepairUnitPrepStmt.bind(id));
-      }
-    }
-    session.executeAsync(deleteClusterPrepStmt.bind(clusterName));
-    return Optional.ofNullable(new Cluster(clusterName, null, null));
+    assert getRepairRunsWithState(RepairRun.RunState.RUNNING).stream()
```

Comment:
That check/call is done in the ClusterResource method already, twice in case there are matching repairs. Since listing repair runs by state can be quite expensive in Cassandra, we might want to limit the number of calls.

Reply:
It's intentional to be in both places. At the resource level it is required so to provide correct response handling to the user, ie application level. At the storage level it is provided for program correctness. If performance is a problem in production systems it is intended that assertions can be disabled.